### ISSUE TITLE
greedy vs optimal peak-picking?

### ISSUE THREAD

**Issue by bmcfee** [librosa/librosa#164]
(cc @craffel @dpwe)

The current peak picker uses a greedy heuristic to select peaks from a time series `x(t)`, after narrowing down to a set of time points that satisfy some selection criteria (local maxima, significantly larger than local mean, well-separated).

I think we can do better by more carefully defining an optimization problem to be solved.

The greedy algorithm is essentially finding a feasible solution to a set of constraints, but otherwise has no selection criterion.  However, because of the one-dimensional structure of the problem, a dynamic programming solution falls out naturally if we define an objective function.

Going forward in time from the first feasible peak (say, `t`), we have two options: take it or not.  If we take it, we add to the score some value `f(t)` and then we have to wait some number of steps before picking up again at `t + wait`.  Otherwise, we leave the score as is, and pick up at `t+1` (or wherever the next peak lies).  Like the current greedy algorithm, this should be computable in linear time, but with the added benefit that the solution will be in optimal in a specific sense.

The score for a solution is then `sum_t f(t)` over the chosen times `t`, and `f(t)` defines a point-wise score function.  Some possible choices for the score include:
- `f(t) = 1` : maximize the number of peaks chosen
- `f(t) = x(t)` : prefer large values (eg, onset strength) for the selected peaks
- others?

**Comment by bmcfee** [comment]
> The constraint is bi-directional. I can't decide whether to include this
> peak until I see what comes after: is it a local max (ok, 1 step fixed
> delay), but also is there a bigger, better peak coming after?

Right, this would be the point of using a DP.  The recurrence for the value at time `t` would look like:
```
value(t) = max(f(t) + value(t + wait), value(t + 1))
```
A peak is either selected and inhibits those immediately after it (first cast), or not selected, and thus inherits the best value from the following timestep.  This optimization does not need to be bi-directional.

**Comment by ebattenberg** [comment]
Are there different peak picking modes?  I ask because in most of my work,
I've only used causal peak-picking (with a small amount of latency) because
of the real-time aspect of the applications.

I guess the algorithm you just proposed is causal during the first pass.
But is there then a non-causal backtracking/decoding step?

**Comment by bmcfee** [comment]
> Are there different peak picking modes?  I ask because in most of my work,
> I've only used causal peak-picking (with a small amount of latency) because
> of the real-time aspect of the applications.

The DP would actually be a single backward pass, so it would not be real-time friendly.

Note: I'm not suggesting to replace the existing picker, but to supplement it with the DP solution.

**Comment by dpwe** [comment]
I don't get it.  My intuition for DP is that I work up to time T, and at
that time I can figure out whether to include or not include the peak at
that point, and that I won't have to revisit that decision later on.

Your recurrence appears to be DP working backwards in time, but the peak
picking process, I believe, should be symmetric in time (we're just looking
for prominent peaks, we're not talking about some intrinsically causal
constraint).  So I should be able to make sense of it working forwards in
time, too:  score(t) = max(local_score(t) + score(t - wait), score(t - 1)).

Hmm, it still doesn't feel like DP.  But in principal, this sounds like the
dp beat tracker ... score(t) = max_{\tau : \tau > wait} (score(t - \tau)) +
local_score(t)

But I can't figure how to cast the direction-symmetric peak picking problem
into a DP form.

  DAn.

**Comment by bmcfee** [comment]
> Your recurrence appears to be DP working backwards in time, but the peak
> picking process, I believe, should be symmetric in time (we're just looking
> for prominent peaks, we're not talking about some intrinsically causal
> constraint).

No, but if we're breaking the problem into smaller pieces, we do need to settle on a single direction.  The reverse would also work, as you mention, but it feels less natural to me.

In either case, it still can't work online as it needs to recurse all the way down to the base case (end of the signal/empty candidate set).

**Comment by bmcfee** [comment]
Following up on this after a bazillion years.

I hacked a DP solver for this a while back, and it works fine.  In most examples, I see zero qualitative difference relative to the greedy solver - at most the results are off by a frame or two.

What is interesting here is that the DP solver is trying to maximize the aggregate spectral flux of selected peaks.  Whenever the DP differs from greedy, it's always in the direction of coming *later* in the event, eg at the peak of the ADSR envelope.  This is actually in direct opposition to the idea of onset backtracking, where detections roll back to the beginning of ADSR.  I haven't benchmarked this quantitatively, but my hunch is that having onsets land earlier than the peak is generally a good thing.  Of course, we always have the option of rolling back the DP-selected peaks as a post-processing step.

At this point the only thing the DP really has going for it is consistency; its behavior is unambiguous and globally optimal, where the greedy solver is not.  How much that matters is questionable at best.

<details>

<summary> refactored onset code with greedy and DP solvers </summary>

```python
def peak_pick(x, pre_max, post_max, pre_avg, post_avg, delta, wait):

    if pre_max < 0:
        raise ParameterError("pre_max must be non-negative")
    if pre_avg < 0:
        raise ParameterError("pre_avg must be non-negative")
    if delta < 0:
        raise ParameterError("delta must be non-negative")
    if wait < 0:
        raise ParameterError("wait must be non-negative")

    if post_max <= 0:
        raise ParameterError("post_max must be positive")

    if post_avg <= 0:
        raise ParameterError("post_avg must be positive")

    if x.ndim != 1:
        raise ParameterError("input array must be one-dimensional")

    # Ensure valid index types
    pre_max = valid_int(pre_max, cast=np.ceil)
    post_max = valid_int(post_max, cast=np.ceil)
    pre_avg = valid_int(pre_avg, cast=np.ceil)
    post_avg = valid_int(post_avg, cast=np.ceil)
    wait = valid_int(wait, cast=np.ceil)

    # Get the maximum of the signal over a sliding window
    max_length = pre_max + post_max
    max_origin = np.ceil(0.5 * (pre_max - post_max))
    # Using mode='constant' and cval=x.min() effectively truncates
    # the sliding window at the boundaries
    mov_max = scipy.ndimage.filters.maximum_filter1d(
        x, int(max_length), mode="constant", origin=int(max_origin), cval=x.min()
    )

    # Get the mean of the signal over a sliding window
    avg_length = pre_avg + post_avg
    avg_origin = np.ceil(0.5 * (pre_avg - post_avg))
    # Here, there is no mode which results in the behavior we want,
    # so we'll correct below.
    mov_avg = scipy.ndimage.filters.uniform_filter1d(
        x, int(avg_length), mode="nearest", origin=int(avg_origin)
    )

    # Correct sliding average at the beginning
    n = 0
    # Only need to correct in the range where the window needs to be truncated
    while n - pre_avg < 0 and n < x.shape[0]:
        # This just explicitly does mean(x[n - pre_avg:n + post_avg])
        # with truncation
        start = n - pre_avg
        start = start if start > 0 else 0
        mov_avg[n] = np.mean(x[start : n + post_avg])
        n += 1
    # Correct sliding average at the end
    n = x.shape[0] - post_avg
    # When post_avg > x.shape[0] (weird case), reset to 0
    n = n if n > 0 else 0
    while n < x.shape[0]:
        start = max(n - pre_avg, 0)
        
        mov_avg[n] = np.mean(x[start : n + post_avg])
        n += 1

    # First mask out all entries not equal to the local max
    detections = x * (x == mov_max)

    # Then mask out all entries less than the thresholded average
    detections = detections * (detections >= (mov_avg + delta))

    return detections, wait
    #return greedy_select(detections, wait)


def greedy_select(detections, wait):
    # Initialize peaks array, to be filled greedily
    peaks = []

    # Remove onsets which are close together in time
    last_onset = -np.inf

    for i in np.nonzero(detections)[0]:
        # Only report an onset if the "wait" samples was reported
        if i > last_onset + wait:
            peaks.append(i)
            # Save last reported onset
            last_onset = i

    peaks = np.array(peaks)
    
    return peaks


def dp_select(x, wait):
    
    values = np.zeros(len(x)+1)
    pointers = np.zeros(len(x)+1, dtype=int)
    taken = np.zeros_like(values, dtype=bool)
    
    values[-1] = 0
    pointers[-1] = -1
    
    for n in range(len(x)-1, -1, -1):
        next_ptr = min(len(x), n + wait)
        
        if x[n] <= 0:
            # If x[n] is not a peak, no problem
            values[n] = values[n+1]
            pointers[n] = n+1
            
        elif x[n] + values[next_ptr] > values[n+1]:
            # x[n] is a peak, but is it better to take it and skip the next [wait]?
            values[n] = x[n] + values[next_ptr]
            pointers[n] = next_ptr
            taken[n] = True
        else:
            # don't take x[n], keep looking
            values[n] = values[n+1]
            pointers[n] = n+1
          
    selections = []
    n = 0
    while pointers[n] >= 0:
        if taken[n]:
            selections.append(n)
        n = pointers[n]
        
    selections = np.array(selections)
    
    return selections
```

</details>

**Comment by bmcfee** [comment]
Note: the solution above needs to be rewritten a bit in the wake of #1766 but it should be completely doable.
